src/main/java/me/jellysquid/mods/hydrogen/common/state/property/WeakPropertyCache.java

```diff
@@ -0,0 +1,26 @@
+package me.jellysquid.mods.hydrogen.common.state.property;
+
+import net.minecraft.state.property.Property;
+
+import java.lang.ref.WeakReference;
+import java.util.WeakHashMap;
+
+public class WeakPropertyCache {
+
+    /**
+     * Values are {@link WeakReference} because values in {@link WeakHashMap} are strongly-referenced by default.
+     */
+    private static final WeakHashMap<Property<?>, WeakReference<Property<?>>> CACHE = new WeakHashMap<>();
+
+    /**
+     * Caches properties passed into it in a {@link WeakHashMap}.
+     *
+     * @param newProperty the property to cache
+     * @param <T>         the property type (checked by the property's equals method)
+     * @return the cached property
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends Property<?>> T tryCacheProperty(T newProperty) {
+        return (T) CACHE.computeIfAbsent(newProperty, WeakReference::new).get();
```

Comment:
The WeakReference::get() can probably return null if it is collected after computeIfAbsent returns but before get() is executed. I suggest a local variable and in case of null, put the new property into the map.

Reply:
The weak reference just created with the constructor reference is based on `newProperty`, so it won't be collected until `tryCacheProperty` returns. _However_ if `computeIfAbsent` finds a property already in `CACHE` (that isn't `newProperty`), GC could run between `computeIfAbsent` and `get` and finalize+clear the reference which would cause issues.

Do you think something like this would work?
```java
T cachedProperty = (T) CACHE.computeIfAbsent(newProperty, WeakReference::new).get();
if (cachedProperty == null) {
    CACHE.put(newProperty, new WeakReference<>(newProperty));
    return newProperty;
} else {
    return cachedProperty;
}
```
It kinda smells like double-checked locking, but I think it should work because `newProperty` is hard-referenced throughout the entire method.

Reply:
If that whole thing is still inside the synchronized(CACHE) block, yes I think that will work.